internetarchive/openlibrary
# Add messages.pot update script to pre-commit.

Closes #8818.


Feature. Automates generation of new `messages.pot` file for translation whenever Python or HTML files that affect `i18n` strings are changed.

### Technical

The implementation is in two parts and mainly confined to two files:
1. `pre-commit-config.yaml` -- Added a new `pre-commit` local hook that runs the `./scripts/i18n-messages extract` script to generate a new `messages.pot` file and is activated by changes to Python or HTML files.

**Note:** This hook has a few dependencies will need to be updated to match the current version when we update `Babel`, `multipart`, and/or `web.py`.

2. `__init__.py` in `openlibrary/i18n` -- Wrote new logic into the message extraction function, using a straightforward `symmetric_difference` compare between the old and new `i18n` messages to make sure that `messages.pot` only gets updated when `i18n` strings have been added/modified/deleted. 

**Note:** This was done to prevent an infinite loop of failure within the pre-commit hook, as the hook will otherwise update the template time field in `messages.pot` on every run, and report a failed run as a result of the file modification, which prevents CI autofixing.

This extra logic also had the nice side effect of ensuring that there will be no `messages.pot` updates for totally unrelated files, as only actual relevant string changes will trigger the re-generation.

### Testing

1. Make a change to a Python or HTML file that adds, removes or modifies `i18n`-formatted text
4. Install `pre-commit` if not installed
5. Commit the change; `pre-commit` should automatically generate a new `messages.pot` file

### Stakeholders

@cdrini
